src/hypothesis/searchstrategy/types.py

```diff
@@ -0,0 +1,260 @@
+        fractions.Fraction: st.fractions(),
+        decimal.Decimal: st.decimals(),
+        text_type: st.text(),
+        bytes: st.binary(),
+        dt.datetime: st.datetimes(),
+        dt.date: st.dates(),
+        dt.time: st.times(),
+        dt.timedelta: st.timedeltas(),
+        uuid.UUID: st.uuids(),
+        # Built-in types
+        type: st.sampled_from([type(None), bool, int, str, list, set, dict]),
+        type(Ellipsis): st.just(Ellipsis),
+        type(NotImplemented): st.just(NotImplemented),
+        bytearray: st.binary().map(bytearray),
+        memoryview: st.binary().map(memoryview),
+        # Pull requests with more types welcome!
+    }
+    for t in integer_types:
+        known_type_strats[t] = st.integers()
+    # build empty collections, as only generics know their contents
+    known_type_strats.update({
+        t: st.builds(t) for t in (tuple, list, set, frozenset, dict)
+    })
+    try:
+        from hypothesis.extra.pytz import timezones
+        known_type_strats[dt.tzinfo] = timezones()
+    except ImportError:  # pragma: no cover
+        pass
+    try:  # pragma: no cover
+        import numpy as np
+        from hypothesis.extra.numpy import \
+            arrays, array_shapes, scalar_dtypes, nested_dtypes
+        known_type_strats.update({
+            np.dtype: nested_dtypes(),
+            np.ndarray: arrays(scalar_dtypes(), array_shapes(max_dims=2)),
+        })
+    except ImportError:
+        pass
+    return known_type_strats
+
+
+def type_sorting_key(t):
+    """Minimise to None, then non-container types, then container types."""
+    if t is None or t is type(None):
+        return -1
+    return issubclass(t, collections.abc.Container)
+
+
+def check_by_origin(thing, super_):
+    return thing is super_ or getattr(thing, '__origin__', None) is super_
+
+
+def try_issubclass(thing, maybe_superclass):
+    try:
+        return issubclass(thing, maybe_superclass)
+    except (AttributeError, TypeError):
+        # Some types can't be the subject or object of an instance or
+        # subclass check, including _TypeAlias and Union
+        return False
+
+
+def from_typing_type(thing):
+    import typing
+    # `Any` and `Type` mess up our subclass lookups, so handle them first
+    if thing is typing.Any:
+        raise ResolutionFailed('Cannot resolve typing.Any to any strategy.')
+    if check_by_origin(thing, getattr(typing, 'Type', object())):
+        if thing.__args__ in (None, typing.Any):
+            return st.just(type)
+        return st.from_type(thing.__args__[0]).map(type)
+    if check_by_origin(thing, typing.Union) or \
+            (sys.version_info[:2] == (3, 5) and thing.__name__ == 'Union'):
+        params = getattr(thing, '__union_params__', None) or ()
+        args = getattr(thing, '__args__', None) or ()
+        possible = sorted(params + args, key=type_sorting_key)
+        if not possible:
+            raise ResolutionFailed('Cannot resolve Union of no types.')
```

Comment:
Hilarious. It looks like this is another version difference.

```bash
$ pyenv shell 3.6.1
$ python -c 'import typing; print(isinstance(typing.Union, type))'
False
$ pyenv shell 3.6.0
$ python -c 'import typing; print(isinstance(typing.Union, type))'
False
$ pyenv shell 3.5.1
$ python -c 'import typing; print(isinstance(typing.Union, type))'
True
```

My inclination is to make an empty Union an error on Python 3.5 as well given the fact that in more recent versions it will be if we block non-types.